asmcomp/power/emit.mlp

```diff
@@ -699,20 +726,8 @@ let emit_instr env i =
             emit_set_comp c i.res.(0)
         end
     | Lop(Icompf cmp) ->
-        ignore cmp;
-        Misc.fatal_error "Icompf is not implemented for this platform";
-        (* smuenzel: the following code may be used to implement
-           Icompf when this platform is supported again post 5.0.
-           It will need to be tested.
-           {[
-             let res = i.res.(0) in
-             let bitnum, negated = emit_float_comp cmp i.arg in
-             `	mfcr	0\n`;
-             `	rlwinm	{emit_reg res}, 0, {emit_int(bitnum+1)}, 31, 31\n`;
-             if negated
-             then `  xori  {emit_reg res}, {emit_reg res}, 1\n`
-           ]}
-        *)
+        let (bitnum, negated) = emit_float_comp cmp i.arg in
+        emit_extract_crbit bitnum negated i.res.(0)
```

Comment:
The memory model is implemented by the following sequences (compared to [RISC-V](https://github.com/ocaml/ocaml/pull/11418) and [ARMv8](https://github.com/ocaml/ocaml/pull/10972)):
| OCaml Operation |             POWER operation |                 ARMv8 operation |                       RISC-V operation |
|:----------------|----------------------------:|--------------------------------:|---------------------------------------:|
| Atomic load     | sync; lwz; cmpw; bne; isync |                 dmb ishld; ldar | fence iorw, iorw; ld; fence iorw, iorw |
| Atomic store    |                 lwsync; stw | L: ldaxr; stlxr; cbnz L; dmb st |           fence iorw, ow; amoswap.d.aq |
| Nonatomic load  |                         lwz |                             ldr |                                     ld |
| Nonatomic store |                         stw |                  dmb ishld; str |                                     sd |

These look right to me and match the instructions I had worked out. 

It uses these instructions:
 * sync (hwsync) - heavy weight sync is a full memory barrier that prevents any reordering
 * lwsync - lightweight sync is a memory barrier that prevents reordering (Load-Load, Load-Store, Store-Store)
 * isync - instruction sync, refetches any instructions that might have been fetched prior to this instruction
 
 I'm running this through the [ocaml-multicore/multicoretests](https://github.com/ocaml-multicore/multicoretests) to validate further.

Reply:
[ocaml-multicore/multicoretests](https://github.com/ocaml-multicore/multicoretests) didn't find any new issues (that aren't already reported elsewhere). I've compiled various EIO/multicore projects and no problems found. 

I'd like to setup https://check.ci.ocaml.org to run against CI's POWER pool and see what it finds. @avsm

Reply:
Non-initialising (non-atomic) stores for Power are incorrect in the table. The PR does the right thing and emits: `lwsync; stw`, which is the same as what the PLDI 2018 paper suggests. See sec 8.2 in the paper: https://kcsrk.info/papers/pldi18-memory.pdf. In the code, see https://github.com/ocaml/ocaml/pull/12276/files#diff-f99d939a29505721a1d19a6aed606eac2d97eff24a8936102ab32e0794a3a97aR695.